### PR TITLE
Update podman image module to support docker-archive transport cleanly

### LINKED ISSUE #205: Podman Image task doesn't really support saving to disk
/kind bug

**Description**

The podman image tasks looks like it should support exporting a container image to a tar ball. Looking at the documentation one would use the push options, combined with the transport as "docker-archive". In practice this doesn't work because the regex code in the plugins/modules/podman_image.py can't really handle it. It's difficult to explain, you kind of have to try it and see the issues. I'll include the examples and what happens below.

**Steps to reproduce the issue:**

1. Write a task to save a contianer image to disk

2. Run ansible

**Describe the results you received:**

Various error messages based on how I contstructed the task

**Describe the results you expected:**

A tar ball of the container image

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `ansible --version`:**

```
ansible 2.9.17
  config file = /opt/data/danclark/workspace/openshift-disconnected/playbooks/ansible.cfg
  configured module search path = ['/home/danclark/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  executable location = /usr/bin/ansible
  python version = 3.9.1 (default, Jan 20 2021, 00:00:00) [GCC 10.2.1 20201125 (Red Hat 10.2.1-9)]
```

**Output of `podman version`:**

```
Version:      2.2.1
API Version:  2.1.0
Go Version:   go1.15.5
Built:        Tue Dec  8 09:37:50 2020
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
Version:      2.2.1
API Version:  2.1.0
Go Version:   go1.15.5
Built:        Tue Dec  8 09:37:50 2020
OS/Arch:      linux/amd64
[danclark@localhost playbooks]$ podman info --debug
host:
  arch: amd64
  buildahVersion: 1.18.0
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: conmon-2.0.26-1.fc33.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.26, commit: 777074ecdb5e883b9bec233f3630c5e7fa37d521'
  cpus: 8
  distribution:
    distribution: fedora
    version: "33"
  eventLogger: journald
  hostname: localhost.localdomain
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65536
  kernel: 5.10.16-200.fc33.x86_64
  linkmode: dynamic
  memFree: 23347650560
  memTotal: 33555304448
  ociRuntime:
    name: crun
    package: crun-0.17-1.fc33.x86_64
    path: /usr/bin/crun
    version: |-
      crun version 0.17
      commit: 0e9229ae34caaebcb86f1fde18de3acaf18c6d9a
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +CRIU +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.8-1.fc33.x86_64
    version: |-
      slirp4netns version 1.1.8
      commit: d361001f495417b880f20329121e3aa431a8f90f
      libslirp: 4.3.1
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.0
  swapFree: 21147672576
  swapTotal: 21147672576
  uptime: 11h 13m 12.5s (Approximately 0.46 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /home/danclark/.config/containers/storage.conf
  containerStore:
    number: 0
    paused: 0
    running: 0
    stopped: 0
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.4.0-1.fc33.x86_64
      Version: |-
        fusermount3 version: 3.9.3
        fuse-overlayfs: version 1.4
        FUSE library version 3.9.3
        using FUSE kernel interface version 7.31
  graphRoot: /home/danclark/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 4
  runRoot: /run/user/1000/containers
  volumePath: /home/danclark/.local/share/containers/storage/volumes
version:
  APIVersion: 2.1.0
  Built: 1607438270
  BuiltTime: Tue Dec  8 09:37:50 2020
  GitCommit: ""
  GoVersion: go1.15.5
  OsArch: linux/amd64
  Version: 2.2.1
```

First example tasks:
```
- name: pull image
  containers.podman.podman_image:
    name: 'registry.redhat.io/openshift3/ose-ansible'
    tag: 'v3.11'
    push: true
    #pull: false
    push_args:
      #compress: true
      dest: '/opt/data/openshift/ocp-3.11.318/images/infrastructure'
      transport: 'docker-archive'
```

This tasks results in a podman command that looks like this:
```
podman push registry.redhat.io/openshift3/ose-ansible:v3.11
```
That's clearly not going to work. What if I try
```
- name: pull image
  containers.podman.podman_image:
    name: 'registry'
    tag: '2'
    push: true
    #pull: false
    push_args:
      #compress: true
      dest: '/opt/data/openshift/images/infrastructure'
      transport: 'docker-archive'
```
That results in a podman command that looks like this
```
podman push registry:2 docker-archive:/opt/data/openshift/images/infrastructure
```

And the ansible output:

```
TASK [mirror_ocp_images : pull image] ********************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {
    "changed": false
}

MSG:

Failed to push image registry:2: Error: error copying image to the remote destination: Error initializing destination docker-archive:/opt/data/openshift/images/infrastructure: error opening file "/opt/data/openshift/images/infrastructure": open /opt/data/openshift/images/infrastructure: is a directory
```

How about this:
```
- name: pull image
  containers.podman.podman_image:
    name: 'registry'
    tag: '2'
    push: true
    #pull: false
    push_args:
      #compress: true
      dest: '/opt/data/openshift/images/infrastructure/registry.tar'
      transport: 'docker-archive'
```

While results in a podman command like this:
```
podman push registry:2 docker-archive:/opt/data/openshift/images/infrastructure.tar
```

The only way I can get the image save to work is:
```
- name: pull image
  containers.podman.podman_image:
    name: 'registry'
    tag: '2'
    push: true
    #pull: false
    push_args:
      #compress: true
      dest: 'registry.tar'
      transport: 'docker-archive'
```

That will create a registry.tar file in the local working directory. But if I use the full name of the image:
```
- name: pull image
  containers.podman.podman_image:
    name: 'docker.io/library/registry'
    tag: '2'
    push: true
    #pull: false
    push_args:
      #compress: true
      dest: 'registry.tar'
      transport: 'docker-archive'
```

Ansible tries to push it to docker.io which won't work
```
TASK [mirror_ocp_images : pull image] ********************************************************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => {
    "changed": false
}

MSG:

Failed to push image docker.io/library/registry:2: Getting image source signatures
Copying blob sha256:4f5aa08c5eaa01cf2d4f940eca389e3ef597af52836061640a46ce29b52bd18b
Copying blob sha256:8ebb9d6ed165ab026689173ee83d23eedfe8dd8128fd2dc5c0765f235e1f302d
Copying blob sha256:fb6b1a93008f9c70cded57e5469abfd34c383c053fd14a58ff5546ee287f5fe5
Copying blob sha256:6d2d8cb41f01e47d5959e4606e054fad6496d3d7711c270c446f3c6a83ff514b
Copying blob sha256:0fcbbeeeb0d7fc5c06362d7a6717b999e605574c7210eff4f7418f6e9be9fbfe
Error: error copying image to the remote destination: Error writing blob: Error initiating layer upload to /v2/library/registry/blobs/uploads/ in registry-1.docker.io: errors:
denied: requested access to the resource is denied
unauthorized: authentication required
```

It sort of looks like the docker-archive functionality was added but not really flushed out. The if/then cases and regex patterns in the code cause all sorts of weird cases.

### PR DESCRIPTION
**Pull request by dmc5179**
Current podman image module doesn't really support the docker-archive transport. This update allows the docker-archive transport to work as expected.

Fix #205

### CODE REVIEW

**Comment by sshnaidm** [comment]
Would you like to add tests here? https://github.com/containers/ansible-podman-collections/blob/master/tests/integration/targets/podman_image/tasks/main.yml

**Comment by dmc5179** [comment]
@sshnaidm Tests added and fix simplified.

File: plugins/modules/podman_image.py [hunk 1]
```diff
@@ -667,7 +668,7 @@ def push_image(self):
         dest_string = dest_format_string.format(transport=transport, name=self.name, dest=dest, image_name=self.image_name,)
 
         # Only append the destination argument if the image name is not a URL
```
Comment (sshnaidm) [inline plugins/modules/podman_image.py:670]: this comment is not true now, right?
` # Only append the destination argument if the image name is not a URL`

File: tests/integration/targets/podman_image/tasks/main.yml [hunk 2]
```diff
@@ -40,6 +59,48 @@
...
+    - name: Push to a local directory
+      containers.podman.podman_image:
+        name: 'docker.io/library/alpine'
+        push: true
+        push_args:
+          dest: '/tmp/alpine'
+          transport: 'dir'
+
+    - name: Remove local archive testing directory (post-test)
+      ansible.builtin.file:
```
Comment (sshnaidm) [inline tests/integration/targets/podman_image/tasks/main.yml:96]: Can we check that `/tmp/alpine` exists before deleting? Like it's done above for `register: alpine_docker_archive`

**Review by sshnaidm — CHANGES_REQUESTED** [review]
Left comments inside, thanks!

File: plugins/modules/podman_image.py [hunk 3]
```diff
@@ -667,7 +668,7 @@ def push_image(self):
...
-        if '/' not in self.name:
+        if transport or transport == 'docker':
```
[the remark this answers is not in the document]
Comment (sshnaidm) [reply plugins/modules/podman_image.py:671]: Now I think - maybe you meant `and` instead of `or`?

**Comment by ocafebabe** [comment]
Any updates regarding this PR?

**Comment by sshnaidm** [comment]
So far I released 1.8.0, this fix will be in 1.8.1

File: tests/integration/targets/podman_image/tasks/main.yml (hunk 2 above)
Comment (dmc5179) [reply tests/integration/targets/podman_image/tasks/main.yml:96]: I cleaned up the tests. I think I wrote them too late at night :)

File: plugins/modules/podman_image.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (dmc5179) [reply plugins/modules/podman_image.py:671]: The comment on 670 is old. There was a reason I made that crazy or statement but looking at it now it doesn't make sense. Need to look at it again.

File: tests/integration/targets/podman_image/tasks/main.yml [hunk 4]
```diff
@@ -29,6 +29,42 @@
...
+      stat:
+        path: '/tmp/alpine.tar'
+      register: alpine_docker_archive
+
+    - name: Remove OCI format tarball (pre-test)
+      ansible.builtin.file:
+        path: '/tmp/alpine_oci.tar'
+        state: absent
+
+    - name: Push to a tarball in the OCI format
```
Comment (sshnaidm) [inline tests/integration/targets/podman_image/tasks/main.yml:55]: this fails in CI with message:
```
Failed to run ['/usr/bin/podman', 'inspect', '696d33ca1510966c426bdcc0daf05f75990d68c4eb820f615edccf7b971935e7', '--format', 'json'] ['inspect', '696d33ca1510966c426bdcc0daf05f75990d68c4eb820f615edccf7b971935e7', '--format', 'json']: Error: error inspecting object: no such object: "696d33ca1510966c426bdcc0daf05f75990d68c4eb820f615edccf7b971935e7"
```

**Comment by sshnaidm** [comment]
@dmc5179 you need to rebase on latest, I pushed some fixes for CI recently.

**Comment by ocafebabe** [comment]
Any updates regarding this issue?

**Comment by sshnaidm** [comment]
please see https://github.com/containers/ansible-podman-collections/pull/619

**Pull request closed without merging**